### DISCUSSION TITLE
rescue mode (finit,initctl and (reboot etc) symlinks not working as expected

### DISCUSSION THREAD

**Discussion by hongkongkiwi** [finit-project/finit#243] [General]
So I jumped into rescue mode to test it out (I have sulogin on my system) it seems the internal finit communication doesn't work properly (is this intentional?)

I noticed it as I like to type (reboot):
```
# reboot
reboot: Failed connecting to finit: No such file or directory
# ls -lh /sbin/reboot
lrwxrwxrwx    1 root     root           7 Apr  7  2021 /sbin/reboot -> initctl
# which reboot
/sbin/reboot
```

Same thing happens with initctl
```
# initctl status
initctl: Failed connecting to finit: No such file or directory
# which initctl
/sbin/initctl
# ls -lh /sbin/initctl
-rwxr-xr-x    1 root     root       49.6K Apr  7  2021 /sbin/initctl
```

Also finit itself:
```
# finit 6
initctl: Failed connecting to finit: No such file or directory
```

Using the `exit` command works fine to continue to boot the system as normal (presumably because it's a builtin shell command)

**Comment by hongkongkiwi** [comment]
Ah, and here is a related issue I think.... rescue booting and exiting is a bit inconsistent, I can replicate some funny issues around this. I am using the sulogin rescue rather than finit rescue conf.

Booting rescue mode (using `-- rescue` in setargs) drops me immediately into this text without displaying any banner:
```
Uncompressing Linux... done, booting the kernel.
Give root password for maintenance (Ctrl-D to continue):
```

Ok, then after this, if I just type `exit` I get one of three different possibilities (seems randomly):
1. Immediately drops me back to this: `Give root password for maintenance (Ctrl-D to continue):`
2. Starts a half hearted bootup with banner:
```
● ● ●  CustomBanner ══════════════════════════════════════════════════════
[ OK ] Mounting filesystems
[ OK ] Enabling system swap
[ OK ] Cold plugging system
[ OK ] Entering rescue mode
Give root password for maintenance (Ctrl-D to continue):
```
3. Attempts a full boot:
```
● ● ●  CustomBanner ══════════════════════════════════════════════════════
[ OK ] Mounting filesystems
[ OK ] Enabling system swap
[ OK ] Cold plugging system
[ OK ] Initializing random number generator
[ OK ] Starting System log daemon
[ OK ] Starting Kernel log daemon
[ OK ] Starting Hotplug daemon
[ OK ] Device Coldplug one-shot
[ OK ] Bringing up network interfaces ...
<snip>
```

This seems to be affected on whether I just immediately type "exit" or I try to do something like finit 6 or reboot (from the above comment), then type exit.

I think there's some states around this which are not quite updated. It's not an important issue as rescue mode will be hardly used, but definately a bug.

**Reply by troglobit** [reply]
Ah, yeah this needs to be reviewed again.  Thank you for bringing this to my attention!

In short, reboot + initctl communicate with Finit over a UNIX socket in /var/run/finit/sock, but /var/run (or /run) is likely not set up at the point of sulogin.  The reboot (initctl) tool should detect this case and revert to send signals to PID 1 instead.  

Not sure why exit sometimes triggers another sulogin, that should only happen if sulogin returns non-zero exit code.  Need to look into that, because that should of course work -- exit after `-- rescue` should continue with normal boot, whereas exit from sulogin after failed mount/fsck should trigger a reboot.  (I'll see to updating the documentation with this as well.)
